source/godang_ws/src/vision/vision/depth_scale_ball.py

```diff
@@ -0,0 +1,150 @@
+
+import numpy as np
+from ultralytics import YOLO
+import cv2
+# Load a model
+model = YOLO("../ML_ball/best_yolov8.pt")  # load an official model
+# model = YOLO("path/to/best.pt")  # load a custom trained model
+
+# Export the model
+# model.export(format="onnx")
+frame = cv2.imread("../ML_ball/ball_images/frame_0225.jpg")
+
+def UndistortImg(img):
+  # data from calibration
+  mtx = np.matrix([[1.02896097e+03, 0, 1.01324017e+03], [0.0, 9.92389966e+02, 5.48550898e+02],[0, 0,  1]])
+  dist = np.array([ 0.19576717, -0.2477706,  -0.00620366,  0.00395638,  0.10295289])
+  newcameramtx = np.matrix([[1.07476421e+03, 0, 1.02262547e+03], [0, 1.02925677e+03, 5.43286518e+02],[0, 0, 1]])
+  roi = [13, 14, 1895, 1057]
+
+  dst = cv2.undistort(img, mtx, dist, None, newcameramtx) 
+  # crop the image
+  x, y, w, h = roi
+  return dst[y:y+h, x:x+w]    
+
+
+frame = UndistortImg(frame)
+# print(f"frame Original image shape: {frame.shape}")
+
+print(f"dst Original image shape: {frame.shape}")
+
+results = model(frame) 
+class_names = ['purple','red']
+
+detections = []
+for bbox in results:
+    boxes = bbox.boxes
+    boxes = bbox.boxes
+    cls = boxes.cls.tolist()
+    xyxy = boxes.xyxy
+    conf = boxes.conf
+    masks = bbox.masks
+    for i, class_index in enumerate(cls):
+        class_name = class_names[int(class_index)]
+        confidence = conf[i]
+        x, y, w, h = map(int, xyxy[i])
+        if class_name == 'red':
+            detections.append([x, y, w, h])
+#             cv2.rectangle(frame, (x, y), (w, h), (0, 255, 0), 2)
+#             cv2.imshow('frame', frame)
+#             cv2.waitKey(500)
+# cv2.destroyAllWindows()
+
+
+
+print(detections)
+def calculate_depth(focal_length, real_diameter, bounding_box_width):
+    depth = []
+    if bounding_box_width == 0:
+        raise ValueError("Bounding box width cannot be zero")
+    
+    
+    min_depth = (real_diameter * focal_length) / max(bounding_box_width)
+
+    return min_depth
+
+focal_length_x = 1029.138061543091  
+focal_length_y = 992.6178560916601 
+real_diameter = 0.19
+bounding_box_width = []
+for detection in detections:
+    bounding_box_width.append(detection[2] - detection[0])
+print(f"Bounding box width: {bounding_box_width}")
+
+depth_x = calculate_depth(focal_length_x, real_diameter, bounding_box_width)
+depth_y = calculate_depth(focal_length_y, real_diameter, bounding_box_width)
```

Comment:
สงสัยว่าทำไมใช้แค่ width ทั้ง x และ y

Reply:
```
bounding_box_height = [detection[3] - detection[1] for detection in detections]
``` 
หนูแก้ไขให้คำนวณ depth_y ด้วย bounding_box_height เรียบร้อยค่ะอาจารย์

<details><summary>Details</summary>
<p>

The estimated depth of the ball from the camera (using focal length x) is 2.33 meters.
The estimated depth of the ball from the camera (using focal length y) is 2.12 meters.
Real-world coordinates: X = 1.09 m, Y = 0.27 m, Z = 2.20 m
Nearest bounding box: [1208, 696, 1292, 785]

</p>
</details>